### PR TITLE
Prefer @Nullable to Optional for InterceptingExecutableInvoker

### PR DESCRIPTION
**Pull request by mpkorstanje**
Same as https://github.com/junit-team/junit-framework/pull/5700, but for the `InterceptingExecutableInvoker` specifically. 

An interesting aspect of this API was that we open taken in nullable values, wrap them in an `Optional`, pass them down the call chain and then unwrap them later on. With JSpecify available this isn't necessary at all and we can also remove this behaviour.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit-framework/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://docs.junit.org/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit-framework/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://docs.junit.org/snapshot/) and [Release Notes](https://docs.junit.org/snapshot/release-notes.html)

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ReflectiveInvocationContext.java [hunk 1]
```diff
@@ -56,7 +57,7 @@ public interface ReflectiveInvocationContext<T extends Executable> {
 	 * @return the arguments of the executable in this invocation context;
 	 * immutable and never {@code null}
 	 */
-	List<Object> getArguments();
+	List<@Nullable Object> getArguments();
```
Comment (mpkorstanje) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ReflectiveInvocationContext.java:60]: This nullable annotation was missing. A method may be invoked with `null` as on of it's arguments.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/InterceptingExecutableInvoker.java [hunk 2]
```diff
@@ -89,12 +88,9 @@ public void invokeVoid(Method method, @Nullable Object target, ExtensionContext
 			ExtensionContext extensionContext, ExtensionRegistry extensionRegistry,
 			ReflectiveInterceptorCall<Method, T> interceptorCall) {
 
-		@SuppressWarnings({ "unchecked", "rawtypes" })
-		Optional<Object> optionalTarget = (target instanceof Optional optional ? optional
-				: Optional.ofNullable(target));
```
Comment (mpkorstanje) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/InterceptingExecutableInvoker.java:94]: I couldn't figure out what this is for. It looks like at some point `invoke` could be called with an optional target. I suppose to make it easier to use in combination with `ExtensionContext::getTestInstance`. But that idea never was completed. So I've just removed it.

Comment (sbrannen) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/InterceptingExecutableInvoker.java:94]: Well, if it's no longer used, removing that sounds good.

Please update the `@param target` Javadoc as well.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java [hunk 3]
```diff
@@ -25,23 +25,26 @@
...
 		this.arguments = arguments;
 	}
 
 	@Override
 	public Class<?> getTargetClass() {
-		return this.target.<Class<?>> map(Object::getClass).orElseGet(this.method::getDeclaringClass);
+		if (target != null) {
+			return this.target.getClass();
+		}
+		return this.method.getDeclaringClass();
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java]: We often use a ternary operator for situations like that.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java [hunk 4]
```diff
@@ -50,15 +53,15 @@ public Method getExecutable() {
 	}
 
 	@Override
-	public List<Object> getArguments() {
-		return unmodifiableList(Arrays.asList(this.arguments));
+	public List<@Nullable Object> getArguments() {
+		List<@Nullable Object> list = Arrays.asList(this.arguments);
+		return unmodifiableList(list);
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java:55]: Did you have to extract the `list` in order to appease NullAway?

**Review by sbrannen — APPROVED** [review]

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java (hunk 4 above)
Comment (mpkorstanje) [reply junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java:55]: Yes.

```
> Task :junit-jupiter-engine:compileJava
/home/mpkorstanje/Projects/github/junit-team/junit5/junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/execution/MethodInvocation.java:54: warning: [NullAway] incompatible types: List<Object> cannot be converted to List<@Nullable Object>
                return unmodifiableList(Arrays.asList(this.arguments));
```

**Comment by sormuras** [comment]
> An interesting aspect of this API is that we open taken in nullable values, wrap them in an `Optional`, pass them down the call chain and then unwrap them later on.

So, better store `null`s directly (in fields)?

**Comment by mpkorstanje** [comment]
> So, better store nulls directly (in fields)?

Should have used past tense.: 😛
